fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/IteratorCursorBase.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * IteratorCursorBase.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2019 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.cursors;
+
+import com.apple.foundationdb.API;
+import com.apple.foundationdb.record.ByteArrayContinuation;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.RecordCursor;
+import com.apple.foundationdb.record.RecordCursorResult;
+import com.apple.foundationdb.record.RecordCursorVisitor;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+/**
+ * Base cursor for {@link IteratorCursor} and {@link AsyncIteratorCursor}.
+ * @param <T> the type of elements of the cursor
+ * @param <C> the type of {@link Iterator}
+ */
+@API(API.Status.INTERNAL)
+public class IteratorCursorBase<T, C extends Iterator<T>> implements RecordCursor<T> {
```

Comment:
Should this class be abstract (and maybe package-private?) and then `onNext` is left as an abstract method to be implemented by `IteratorCursor` and `AsyncIteratorCursor`?

Reply:
Yeah, I made it `abstract`.